File: getting_started/selection_sort/selection_sort_python/selection_sort.py
"""
- Implementation of the selection sort algorithm
- Author: Abdoulaye Diallo
- Complexity O(n²)
"""

#def selectionSort(a, n):
#    for i in range(0, n-1):
#        min = i
#        for j in range(i+1, n):
#            if a[j] < a[min]:
#                min = j
#        if min != i:
#            a[i], a[min] = a[min], a[i]        
#    return a



def select(T):
    n = len(T)
    for i in range(n-1):
        min_j = i
        min_x = T[i]
        for j in range(i+1, n):
            if T[j] < min_x:
                min_j = j
                min_x = T[j]
        T[min_j] = T[i]
        T[i] = min_x

def printList(arr):
    for i in range(len(arr)):
        print(arr[i], end=" ")
    print()
    

if __name__ == "__main__":    
    arr = [64, 25, 12, 22, 11]
    print("Given list is ")
    printList(arr)

    select(arr)

    print("Sorted list is ")
    printList(arr)

File: getting_started/selection_sort/selection_sort_cpp/selection_sort.cc
/**
 * Implementation of the selection sort algorithm in C++
 * Author: Abdoulaye Diallo
 * Complexity O(n²) 
*/




#include <iostream>
#include <cstdlib>

/*
void swap(int *xp, int *yp) {
    int  temp = *xp;
    *xp = *yp;
    *yp = temp;
}


void selectionSort(int arr[], int n) {
    int min;
    for (int i = 0; i < n-1; i++) {
        min = i;
        for (int j = i+1; j < n; j++) {
            if (arr[j] < arr[min])
                min = j;  
        }
        if (min != i)
            swap(&arr[i], &arr[min]);
    }
}
*/


void selectionSort(int T[], int n) {
    int min_j, min_x;
    for (int i = 0; i < n-1; i++)
    {
        min_j = i;
        min_x = T[i];
        for (int j = i+1; j < n; j++)
        {
            if (T[j] < min_x)
            {
                min_j = j;
                min_x = T[j];
            }
        }

        T[min_j] = T[i];
        T[i] = min_x;
    }
} 

void printArray(int T[], int n) {
    int i;
    for (i = 0; i < n; i++) {
        std::cout << T[i] << " ";
    }

    std::cout << std::endl;
}


int main(int argc, char const *argv[]) {
    int T[] = {64, 25, 12, 22, 11};
    int n = sizeof(T) / sizeof(T[0]);
    std::cout << "Given array is ";
    printArray(T, n);

    selectionSort(T, n);
    std::cout << "Sorted array is ";
    printArray(T, n);
    return EXIT_SUCCESS;
}
